### PR TITLE
SQL: Avoid empty last pages for GROUP BY queries when possible

### LINKED ISSUE #75528: SQL: queries with GROUP BY always return a scroll cursor in REST API
For the request
```
http://localhost:9200/_sql
Content-Type: application/json

{
    "query": "SELECT gender FROM test_emp GROUP BY gender"
}
```
the response always includes a `cursor` despite the fact that the result set does not contain any more records:
```
{
  "columns": [
    {
      "name": "gender",
      "type": "keyword"
    }
  ],
  "rows": [
    [
      null
    ],
    [
      "F"
    ],
    [
      "M"
    ]
  ],
  "cursor": "u5mzAwFaAWMBCHRlc3RfZW1waQEBCWNvbXBvc2l0ZQdncm91cGJ5AAD/AQAINDA1MWRjOTABBmdlbmRlcgAAAQAA6AcBCgEINDA1MWRjOTAAAU0AAgEAAAAAAP////8PAAAAAAAAAAAAAAAAAVoDAAICAAAAAAAAAAAKAP////8PAQFrCDQwNTFkYzkwAAABAQA="
}
```

Fetching the cursor returns an empty result set:
```
http://localhost:9201/_sql
Content-Type: application/json

{
    "cursor": "u5mzAwFaAWMBCHRlc3RfZW1waQEBCWNvbXBvc2l0ZQdncm91cGJ5AAD/AQAINDA1MWRjOTABBmdlbmRlcgAAAQAA6AcBCgEINDA1MWRjOTAAAU0AAgEAAAAAAP////8PAAAAAAAAAAAAAAAAAVoDAAICAAAAAAAAAAAKAP////8PAQFrCDQwNTFkYzkwAAABAQA="
}
```
```
{
  "rows": []
}
```

Note, if using the `?format=txt` parameter, this reveals a subsequent bug as the empty response cannot be formatted:

```
http://localhost:9201/_sql?format=txt
Content-Type: application/json

{
    "cursor": "u5mzAwFaAWMBCHRlc3RfZW1waQEBCWNvbXBvc2l0ZQdncm91cGJ5AAD/AQAINDA1MWRjOTABBmdlbmRlcgAAAQAA6AcBCgEINDA1MWRjOTAAAU0AAgEAAAAAAP////8PAAAAAAAAAAAAAAAAAVoDAAICAAAAAAAAAAAKAP////8PAQFrCDQwNTFkYzkwAAABAQA="
}
```
```
{
  "error": {
    "root_cause": [
      {
        "type": "sql_illegal_argument_exception",
        "reason": "Cannot find text formatter - this is likely a bug"
      }
    ],
    "type": "sql_illegal_argument_exception",
    "reason": "Cannot find text formatter - this is likely a bug"
  },
  "status": 500
}
```

### PR DESCRIPTION
**Pull request by Luegg**
Resolves https://github.com/elastic/elasticsearch/issues/75528.

Instead of always returning an empty last page for `GROUP BY` queries, `CompositeAggCursor` will now only do so if it is not possible to tell wether there are more pages based on the composite aggregation response. This is the case in two situations:
* The last page contains exactly `fetch_size` results. In this case, the composite aggregation return an `after_key` even if there are no more keys remaining (see also https://github.com/elastic/elasticsearch/issues/75573)
* The query uses a bucket selector. In this case, the composite aggregation might return partial pages with less than `size` buckets and the `buckets.size() < sizeRequested` heuristic for detecting last pages does no longer work.

Hence, if any (or both) of the two conditions above applies, SQL will still return an empty last page. If neither of the conditions apply, the last page will always be non-empty.

This PR is also a weak prerequisite for addressing https://github.com/elastic/elasticsearch/issues/84349 because it allows to immediately close PITs for aggregation queries returning only one page. As a result the performance impact of using PIT for aggregations should be minimized.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java [hunk 1]
```diff
@@ -162,71 +163,64 @@ static void handle(
...
-        // there are some results
-        if (response.getAggregations().asList().isEmpty() == false) {
-            // retry
-            if (shouldRetryDueToEmptyPage(response)) {
-                updateCompositeAfterKey(response, source);
-                retry.run();
-                return;
-            }
 
-            try {
```
Comment (Luegg) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java:180]: The `try/catch` seems to be a leftover that could have been removed in #83833

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java [hunk 2]
```diff
@@ -162,71 +163,64 @@ static void handle(
         BiFunction<SearchSourceBuilder, CompositeAggRowSet, CompositeAggCursor> makeCursor,
         Runnable retry,
         ActionListener<Page> listener,
-        Schema schema
+        boolean mightProducePartialPages
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java:166]: Nit: `couldProducePartialPage` would be more suggestive to me (for both arg and method).

**Review by bpintea — APPROVED** [review]
LGTM.

**Review by luigidellaquila — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java [hunk 3]
```diff
@@ -162,71 +163,64 @@ static void handle(
...
+        Check.isTrue(aggregation instanceof CompositeAggregationBuilder, "Unexpected aggregation builder " + aggregation);
 
-        throw new SqlIllegalArgumentException("Unrecognized root group found; {}", agg.getClass());
+        return (CompositeAggregationBuilder) aggregation;
     }
 
-    private static void updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder search) {
-        CompositeAggregation composite = getComposite(r);
+    static boolean mightProducePartialPages(CompositeAggregationBuilder aggregation) {
+        return aggregation.getPipelineAggregations().stream().anyMatch(a -> a instanceof BucketSelectorPipelineAggregationBuilder);
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java:212]: I don't think we're using any other type of pipeline aggregation so simply returning if the size > 0 should be enough.
Also not a fun of using the streams api due to their cost - the above can become:
```
for (var a : aggregation.getPipelineAggregrations()) {
    return a instanceof BucketSelectorPipelineAggregationBuilder
}
```
```

**Review by costin — APPROVED** [review]
LGTM. A small nit, the issue description could be tweaked a bit to make it clear when the empty page is being consumed (to not pass it to the client) as oppose to when it is NOT, e.g.: when the result size is less than the fetch size AND there are no bucket selector/pipeline aggregations.

Personally I had to read the description several times to understand whether the conditions listed where for the NOT case or not (and thus if there was an AND or OR between them).

**Comment by Luegg** [comment]
Not backporting to 8.1.1 because of changeset overlapping with https://github.com/elastic/elasticsearch/pull/83381

**Pull request merged**